pkg/apis/networking/validation/validation.go

```
@@ -602,3 +602,81 @@ func allowInvalidWildcardHostRule(oldIngress *networking.Ingress) bool {
+func ValidateClusterCIDRConfig(ccc *networking.ClusterCIDRConfig) field.ErrorList {
+	allErrs := apivalidation.ValidateObjectMeta(&ccc.ObjectMeta, false, ValidateClusterCIDRConfigName, field.NewPath("metadata"))
+	allErrs = append(allErrs, ValidateClusterCIDRConfigSpec(&ccc.Spec, field.NewPath("spec"))...)
+	return allErrs
+}
+
+// ValidateClusterCIDRConfigSpec validates clusterCIDRConfig Spec.
+func ValidateClusterCIDRConfigSpec(spec *networking.ClusterCIDRConfigSpec, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if spec.NodeSelector != nil {
+		allErrs = append(allErrs, apivalidation.ValidateNodeSelector(spec.NodeSelector, fldPath.Child("nodeSelector"))...)
+	}
+
+	// Validate if CIDR is configured for at least one IP Family(IPv4/IPv6).
+	if spec.IPv4CIDR == "" && spec.IPv6CIDR == "" {
+		allErrs = append(allErrs, field.Required(fldPath, "one or both of `ipv4` and `ipv6` must be configured"))
+	}
+
+	// Validate configured IPv4 CIDR and PerNodeHostBits.
+	if spec.IPv4CIDR != "" {
+		if !netutils.IsIPv4CIDRString(spec.IPv4CIDR) {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("ipv4CIDR"), spec.IPv4CIDR, "must be a valid IPv4 CIDR"))
+			return allErrs
+		}
+		allErrs = append(allErrs, validatePerNodeHostBits(spec.IPv4CIDR, spec.PerNodeHostBits, 32, fldPath)...)
+	}
+
+	// Validate configured IPv6 CIDR and PerNodeHostBits.
+	if spec.IPv6CIDR != "" {
+		if !netutils.IsIPv6CIDRString(spec.IPv6CIDR) {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("ipv6CIDR"), spec.IPv6CIDR, "must be a valid IPv6 CIDR"))
+			return allErrs
+		}
+		allErrs = append(allErrs, validatePerNodeHostBits(spec.IPv6CIDR, spec.PerNodeHostBits, 128, fldPath)...)
+	}
+
+	return allErrs
+}
+
+func validatePerNodeHostBits(configCIDR string, perNodeHostBits, maxMaskSize int32, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	minPerNodeHostBits := int32(4)
+
+	_, cidr, _ := netutils.ParseCIDRSloppy(configCIDR)
```

we should still check for err here

---

We are checking the error here: https://github.com/sarveshr7/kubernetes/blob/multicidr-rangeallocator/pkg/apis/networking/validation/validation.go#L631 wouldn't that be a redundant error message?

---

I don't know if the linter will complain, in addition, I think that being defensive and check again doesn't harm and protect us of future changes

---

The validation functions called is not exactly the same:

netutils.IsIPv4CIDRString(spec.IPv4CIDR)
netutils.IsIPv6CIDRString(spec.IPv6CIDR)

vs

netutils.ParseCIDRSloppy(configCIDR)


If there is ever a mismatch, we will eat the error.

---

Done